proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+# Package traits
+
+* Proposal: [SE-NNNN](NNNN-swiftpm-package-traits.md)
+* Authors: [Franz Busch](https://github.com/FranzBusch)
+* Review Manager: TBD
+* Status: **Implemented** https://github.com/apple/swift-package-manager/pull/7392
+
+## Introduction
+
+Over the past years the package ecosystem has grown tremendously in both the
+amount of packages and the functionality that individual packages offer.
+Additionally, Swift is being used in more environments such as embedded systems
+or Wasm. This proposal aims to give package authors a new tool to conditionalize
+the features they offer and the dependencies that they use.
+
+## Motivation
+
+There are various motivating use-cases where package authors might want to use traits to
+express conditional compilation or optional dependencies. This section highlights a few.
+
+### Minimizing build times and binary size
+
+Some packages offer different but adjacent functionality such as the
+`swift-collections` package. To reduce the build times and binary size impact
+`swift-collections` offers multiple different products and users can choose
+which one they need. This works however, it comes with the downside that if the
+implementation wants to share code between the different modules it needs to
+create internal targets. Package traits offer another solution to this without
+requiring additional modules.
+
+### Pluggable dependencies
+
+Some packages want to make it configurable what underlying technology is used.
+The [Swift OpenAPIGenerator](https://github.com/apple/swift-openapi-generator)
+for example is capable of running on top of `URLSession`, `AsyncHTTPClient`,
+`Hummingbird` or `Vapor`. To avoid bringing all of those potential dependencies
+into every adopters binary, the project has created individual repositories for
+each transport. This achieves the goal of making the dependencies optional;
+however, it requires users to look at multiple repositories.
+
+### Configurable behavior
+
+Packages often want to cater to multiple ecosystems such as the iOS or the
+server community. While most of the technologies are shared between ecosystems
+there are often some platform specific behaviors/libraries that one might use.
+For example, on Apple's platforms `OSLog` is the canonical logging systems
+whereas the server ecosystem is mostly using `swift-log`. However, there are
+some users that prefer to use `swift-log` on Apple's platforms which means
+libraries and applications cannot use platform compiler conditionals. Package
+traits would allow to make the logging system configurable independent from the
+platform.
+
+### Replacing environment variables in Package manifests
```

Comment:
There's a couple uses of environment variables in the toolchain itself that I don't see covered here:
1. Specifying different package URLs. I don't necessarily think features *should* cover this, but just pointing it out as something else environment variables are used for.
2. Specifying different flags to a target. This one does seem reasonable for features to cover and likely would be with the `.when(traits: ["Foo"])` support, but could be called out specifically?

Reply:
Thanks for bringing this up.
1. Is interesting and we could come up with something where traits can make this work; however, I think the mirror feature or location rewrites are better suited.
2. This is totally in line with what traits can do and I will add it to this proposal